### PR TITLE
Tie decoder embeddings

### PR DESCRIPTION
**Pull request by pmichel31415**
Summary:
Makes it possible to tie the decoder word embeddings and the softmax weights. From my experience this is often helpful especially for small datasets (also this takes less parameters).

Here is a reference for MT (this has been introduced in LM before): https://arxiv.org/abs/1710.01329 (NAACL18)

Differential Revision: D8465257

### CODE REVIEW

**Comment by pmichel31415** [comment]
## Experiments

Here are two experiments with the same configuration except for untied/tied embeddings to show the advantage of this option. 

### Data

This was on the iwslt data provided on the pytorch_translate github (de->en).

### Results

- base (untied):
  - best valid BLEU: 31.30
  - checkpoint size on disk: 453M
- tied:
  - best valid BLEU: 31.61
  - checkpoint size on disk: 374M

### Conclusion

- Small BLEU score gain (+0.31)
- reduces parameter cost by ~17%
---
#### Full model config

```
--log-verbose  \
--no-progress-bar  \
--arch rnn \
--cell-type lstm \
--sequence-lstm  \
--max-tokens 9000 \
--max-sentences 128 \
--max-epoch 999999 \
--stop-time-hr 70 \
--optimizer adam \
--lr 0.001 \
--lr-shrink 0.5 \
--clip-norm 5.0 \
--encoder-dropout-in 0.3 \
--encoder-dropout-out 0.3 \
--decoder-dropout-in 0.3 \
--decoder-dropout-out 0.3 \
--criterion "label_smoothed_cross_entropy" \
--label-smoothing 0.1 \
--encoder-layers 2 \
--encoder-embed-dim 256 \
--encoder-hidden-dim 512 \
--decoder-layers 2 \
--decoder-embed-dim 256 \
--decoder-hidden-dim 512 \
--decoder-out-embed-dim 256 \
--attention-type dot \
--sentence-avg  \
--momentum 0 \
--generate-bleu-eval-per-epoch  \
--beam 5 \
--no-beamable-mm  \
--max-len-a 1.2 \
--max-len-b 10 \
--source-max-vocab-size 40000 \
--target-max-vocab-size 40000 \
--log-interval 500 \
--seed 42 \
```
